addons/base_action_rule/models/base_action_rule.py

```diff
@@ -0,0 +1,396 @@
+        return cron and cron.toggle(model=self._name, domain=[('kind', '=', 'on_time')])
+
+    def _update_registry(self):
+        """ Update the registry after a modification on action rules. """
+        if self.env.registry.ready:
+            # for the sake of simplicity, simply force the registry to reload
+            self.env.cr.commit()
+            self.env.reset()
+            RegistryManager.new(self.env.cr.dbname)
+            RegistryManager.signal_registry_change(self.env.cr.dbname)
+
+    def _get_actions(self, records, kinds):
+        """ Return the actions of the given kinds for records' model. The
+            returned actions' context contain an object to manage processing.
+        """
+        ctx = dict(self.env.context)
+        if '__action_done' not in self.env.context:
+            ctx.update(__action_done={})
+        return self.with_context(ctx, active_test=True).search([('model', '=', records._name), ('kind', 'in', kinds)])
+
+    def _get_eval_context(self):
+        """ Prepare the context used when evaluating python code
+            :returns: dict -- evaluation context given to (safe_)eval
+        """
+        return {
+            'datetime': datetime,
+            'dateutil': dateutil,
+            'time': time,
+            'uid': self.env.uid,
+            'user': self.env.user,
+        }
+
+    def _filter_pre(self, records):
+        """ Filter the records that satisfy the precondition of action ``self``. """
+        if self.filter_pre_id and records:
+            domain = [('id', 'in', records.ids)] + eval(self.filter_pre_id.domain, self._get_eval_context())
+            ctx = eval(self.filter_pre_id.context)
+            return records.with_context(**ctx).search(domain).with_env(records.env)
+        elif self.filter_pre_domain and records:
+            domain = [('id', 'in', records.ids)] + eval(self.filter_pre_domain, self._get_eval_context())
+            return records.search(domain)
+        else:
+            return records
+
+    def _filter_post(self, records):
+        """ Filter the records that satisfy the postcondition of action ``self``. """
+        if self.filter_id and records:
+            domain = [('id', 'in', records.ids)] + eval(self.filter_id.domain, self._get_eval_context())
+            ctx = eval(self.filter_id.context)
+            return records.with_context(**ctx).search(domain).with_env(records.env)
+        elif self.filter_domain and records:
+            domain = [('id', 'in', records.ids)] + eval(self.filter_domain, self._get_eval_context())
+            return records.search(domain)
+        else:
+            return records
+
+    def _process(self, records):
+        """ Process action ``self`` on the ``records`` that have not been done yet. """
+        # filter out the records on which self has already been done, then mark
+        # remaining records as done (to avoid recursive processing)
+        action_done = self.env.context.get('__action_done')
```

Comment:
Please put this line back to

    action_done = self._context['__action_done']

If the key is not in the context, it should crash here.